libfreerdp/core/gateway/rdg.c

```diff
@@ -298,21 +305,113 @@ static BOOL rdg_write_packet(rdpRdg* rdg, wStream* sPacket)
 	return TRUE;
 }
 
-static BOOL rdg_read_all(rdpTls* tls, BYTE* buffer, int size)
+static int rdg_socket_read(BIO* bio, BYTE* pBuffer, size_t size, UINT16 transferEncoding,
+                           size_t* transferEncodingRemainingCount)
+{
+	if (transferEncoding == HTTP_TRANSFER_ENCODING_IDENTITY)
+	{
+		return BIO_read(bio, pBuffer, size);
+	}
+	else if (transferEncoding == HTTP_TRANSFER_ENCODING_CHUNKED)
+	{
+		int status;
+
+		if (transferEncodingRemainingCount == NULL)
+			return -1;
+
+		if (*transferEncodingRemainingCount == 2)
+		{
+			/* footer in buffer. get it now */
+			char _dummy[2];
+			status = BIO_read(bio, _dummy, 2);
+			if (status <= 0)
+			{
+				return status;
+			}
+			if (*transferEncodingRemainingCount >= (size_t)status)
+				*transferEncodingRemainingCount -= (size_t)status;
+			else
+				*transferEncodingRemainingCount = 0;
+		}
+
+		if (*transferEncodingRemainingCount == 0)
+		{
+			char chunkSize[11]; /* ffffffff\r\n\0 should be big enough, else fail. */
+			int _size = 0;
+			BOOL _haveNewLine = FALSE;
+			chunkSize[0] = '\0';
+
+			while (_size < 10 && !_haveNewLine)
```

Comment:
I feel like this function is not supposed to actively wait for a full request to arrive. I may be wrong,  but I feel like you can't be stateless (and BTW you aren't as you have that external `transferEncodingRemainingCount` variable). That means you should probably move `chunkSize` (and `transferEncodingRemainingCount`) in a state variable that would be passed to `rdg_socket_read`. Don't know if I'm clear enough ?

Reply:
ok, but if the state contains 3 things (encoding type, next frame offset and chunk string buffer), maybe it would also make sense creating a struct so that the parameter count of rdg_socket_read does not explode.